Add custom vector configmap setup (#3014)
astronomer/docs: software/export-task-logs.md

:::info

To revert to the default behavior and export task logs using a Fluentd Daemonset, remove this configuration from your `config.yaml` file and reapply it.

:::

#### Customize Vector logging sidecars

You can customize the default Astronomer Vector logging sidecar to have different transformations and sinks based on your team's requirements.

1. Add the following line to your `config.yaml` file:

    ```yaml {5}
    global:
      loggingSidecar:
        enabled: true
        name: sidecar-log-consumer
        customConfig: true
    ```

2. Push the configuration change. See [Apply a config change](apply-platform-config.md).

> 3. Create a custom [vector configuration `yaml` file](https://vector.dev/docs/reference/configuration/) to change how and where sidecars forward your logs. The following examples are template configurations for each commonly used sink:

Author: ```suggestion
3. Create a custom [vector configuration `yaml` file](https://vector.dev/docs/reference/configuration/) to change how and where sidecars forward your logs. The following examples are template configurations for each commonly used sink. For the complete default logging sidecar configmap, see the [Astronomer GitHub](https://github.com/astronomer/airflow-chart/blob/master/templates/logging-sidecar-configmap.yaml).
```

    sources:
      airflow_log_files:
        type: file
        include:
          - "${SIDECAR_LOGS}/*.log"
        read_from: beginning
    transforms:
      transform_syslog:
        type: add_fields
        inputs:
          - airflow_log_files
        fields:
          component: "${COMPONENT:--}"
          workspace: "${WORKSPACE:--}"
          release: "${RELEASE:--}"
    
      transform_task_log:
        type: remap
        inputs:
          - transform_syslog
        source: |-
          # Parse Syslog input. The "!" means that the script should abort on error.
          . = parse_json!(.message)
          .@timestamp = parse_timestamp(.timestamp, "%Y-%m-%dT%H:%M:%S%Z") ?? now()
          .@date_nano = parse_timestamp(.timestamp, "%Y-%m-%dT%H:%M:%S%Z") ?? now()
          .check_log_id = exists(.log_id)
          if .check_log_id != true {
          .log_id = join!([.dag_id, .task_id, .execution_date, .try_number], "_")
          }
          .offset = to_int(now()) * 1000000000 + to_unix_timestamp(now()) * 1000000
    
      transform_remove_fields:
        type: remove_fields
        inputs:
          - transform_task_log
        fields:
          - host

>           - file

Reviewer: https://github.com/astronomer/airflow-chart/blob/master/templates/logging-sidecar-configmap.yaml#L26-L101 - This is our latest config

Author: Do all of the fields in your link have to be configured? I only included this as a suggested template because it was included in Notion.

Author: @pgvishnuram I linked this file above if users want to see the full default configmap, do we also need to update each of these templates with the lines you linked?

Reviewer: Yes we have improved vector data ingestion with these changes for elasticsearch - lets include them

    
    sinks:
      out:

>         type: elasticsearch

Reviewer 2: @jwitz Can we have these as separate example snippets under each sink name
Elastic Search/Datadog etc
Or explicitly mention 'Here is an example for setting ES/ Datadog as a sink'

Author: I added an inline comment at the sinks level, and I also think the tab names will help here.

    ```yaml 
    log_schema:
      timestamp_key : "@timestamp"
    data_dir: "${SIDECAR_LOGS}"
    sources:
      generate_syslog:
        type: file
        include:
          - "${SIDECAR_LOGS}/*.log"
        read_from: beginning
    transforms:
      transform_syslog:
        type: add_fields
        inputs:
          - generate_syslog
        fields:
          component: "${COMPONENT:--}"
          workspace: "${WORKSPACE:--}"
          release: "${RELEASE:--}"
      transform_task_log:
        type: remap
        inputs:
          - transform_syslog
        source: |-
          # Parse Syslog input. The "!" means that the script should abort on error.
          . = parse_json!(.message)
          .@timestamp = parse_timestamp(.timestamp, "%Y-%m-%dT%H:%M:%S%Z") ?? now()
          .check_log_id = exists(.log_id)
          if .check_log_id != true {
          .log_id = join!([.dag_id, .task_id, .execution_date, .try_number], "_")
          }
          .offset = to_int(now()) * 1000000000 + to_unix_timestamp(now()) * 1000000
    sinks:
      my_sink_id:
        type: datadog_logs
        inputs:
          - transform_task_log
        site: us1.datadoghq.com

>         default_api_key: 3a52b4271600d6a012b6789f00794d62

Reviewer: can we remove this api key

>         default_api_key: 3a52b4271600d6a012b6789f00794d62

Author: ```suggestion
        default_api_key: <your-api-key>
```

        encoding:
          codec: json
    ```

  </TabItem>
  <TabItem value="honeycomb">

    ```yaml
    log_schema:
      timestamp_key : "@timestamp"
    data_dir: "${SIDECAR_LOGS}"
    sources:
      generate_syslog:
        type: file
        include:
          - "${SIDECAR_LOGS}/*.log"
        read_from: beginning
    transforms:
      transform_syslog:
        type: add_fields
        inputs:
          - generate_syslog
        fields:
          component: "${COMPONENT:--}"
          workspace: "${WORKSPACE:--}"
          release: "${RELEASE:--}"
    sinks:
      my_sink_id:
        type: honeycomb
        inputs:
          - transform_syslog

>         api_key: nXZwLyMssQZuam9qeMpQiE

Reviewer: can we remove this api key

>         api_key: nXZwLyMssQZuam9qeMpQiE

Author: ```suggestion
        api_key: <your-api-key>
```

        dataset: my-honeycomb-dataset
    
    ```

  </TabItem>

  </Tabs>

4. Run the following command to add the configuration file to your cluster as a Kubernetes secret:

    ```sh

>     kubectl create secret generic sidecar-config --from-file=<your-vector-config>.yaml=<your-vector-config>.yaml

Reviewer: kubectl create secret generic sidecar-config --from-file=vector-config.yaml=vector-config.yaml
customers should not change this name

>     kubectl create secret generic sidecar-config --from-file=<your-vector-config>.yaml=<your-vector-config>.yaml

Author: ```suggestion
    kubectl create secret generic sidecar-config --from-file=vector-config.yaml=vector-config.yaml
```

:::info

To revert to the default behavior and export task logs using a Fluentd Daemonset, remove this configuration from your `config.yaml` file and reapply it.

:::

#### Customize Vector logging sidecars

> You can customize the default Astronomer Vector logging sidecar to have different transformations and sinks based on your team's requirements. This is useful if you want to annotate or otherwise customize or filter your logs before they're sent to your logging platform.

Reviewer 3: ```suggestion
You can customize the default Astronomer Vector logging sidecar to have different transformations and sinks based on your team's requirements. This is useful if you want to annotate, otherwise customize, or filter your logs before sending them to your logging platform.
```

1. Add the following line to your `config.yaml` file:

    ```yaml {5}
    global:
      loggingSidecar:
        enabled: true
        name: sidecar-log-consumer
        customConfig: true
    ```

> 2. Push the configuration change. See [Apply a config change](apply-platform-config.md).

Reviewer 3: Push where? To GH, to the Deployment?

3. Create a custom [vector configuration `yaml` file](https://vector.dev/docs/reference/configuration/) to change how and where sidecars forward your logs. The following examples are template configurations for each commonly used external logging service. For the complete default logging sidecar configmap, see the [Astronomer GitHub](https://github.com/astronomer/airflow-chart/blob/master/templates/logging-sidecar-configmap.yaml).

  <Tabs
      defaultValue="elasticsearch"
      groupId= "customize-vector-logging-sidecars"
      values={[
          {label: 'Elasticsearch', value: 'elasticsearch'},
          {label: 'Honeycomb', value: 'honeycomb'},
          {label: 'Datadog', value: 'datadog'},
      ]}>
  <TabItem value="elasticsearch">

    ```yaml 
    log_schema:
      timestamp_key : "@timestamp"
    data_dir: "${SIDECAR_LOGS}"
    sources:

>       airflow_log_files:

Reviewer 3: Elasticsearch has `aiflow_log-files` as the source while the others both use `generate_syslog` is that expected?
